hal/src/sercom/v2/spi/impl_ehal_thumbv7em.rs

```diff
@@ -0,0 +1,750 @@
+//! Implement [`embedded_hal`] traits for [`Spi`] structs
+//!
+//! As noted in the [spi module](super) documentation, the embedded-hal trait
+//! implementations vary by both [`Size`] and [`Capability`]. Each
+//! implementation is optimized to take advantage of all information known at
+//! compile-time, so it is importatnt to carefully read the documentation in
+//! this module.
+//!
+//! # Variations by [`Size`]
+//!
+//! Remember that SAMx5x chips operate in 32-bit extension mode and use the
+//! hardware `LENGTH` counter to set the number of bytes in each transaction.
+//! The transaction [`Length`] is usually tracked at compile-time using
+//! type-level integers from the [`typenum`] crate, but it can also be tracked
+//! at run-time when using a [`DynLength`].
+//!
+//! The transaction `Length`s can be sub-divided into three groups:
+//! - `Length`s of 1-4 bytes can be completed in a single read/write of the
+//!   `DATA` register. These `Length`s are marked as [`AtomicSize`]s.
+//! - `Length`s [`GreaterThan4`] are known at compile-time but cannot be
+//!   completed atomically.
+//! - A `DynLength` can be any length, but the value is only known at run-time.
+//!
+//! In general, transaction lengths with an `AtomicSize` implement embedded HAL
+//! traits with the corresponding [`Word`] type. For example, [`Spi`] structs
+//! using a transaction `Length` of 2 bytes implement `FullDuplex<u16>`. These
+//! lengths implement both the blocking and non-blocking traits from embedded
+//! HAL. The non-blocking traits are found in the [`spi`] and [`serial`]
+//! modules, while the blocking traits are found in the [`blocking`] module.
+//!
+//! Transaction lengths `GreaterThan4` cannot be completed in a single read or
+//! write of the `DATA` register, so these lengths do **NOT** implement the
+//! non-blocking traits from the embedded HAL `spi` and `serial` modules.
+//! Instead, they only implement traits from the `blocking` module. These traits
+//! are implemented for `u8` types, e.g. `blocking::spi::Transfer<u8>`, and
+//! operate on `[u8]` slices. The length of the slice is checked to ensure it
+//! matches the transaction `Length`.
+//!
+//! Because a `DynLength` is not guaranteed to be an `AtomicSize`, the
+//! corresponding `Spi` structs only implement the `blocking` traits as well.
+//!
+//! For a non-blocking alternative that can be used to transfer arbitrary-length
+//! slices, see the [`spi_future`](super::super::spi_future) module.
```

Comment:
Again, might be nice to mention DMA. (I would advise against feature-gating this behind the `dma` Cargo feature. We want the users to be able to find out that it's a possibility without explicitly having to turn it on beforehand).